### PR TITLE
feat(sweep): Add centralized coordination for parallel sweep workers

### PR DESCRIPTION
**Pull request by akrbc9**
### Major Sweep Infrastructure Refactor
- **Split sweep initialization**: Separated `sweep_init.py` into `sweep_setup.py` (one-time sweep creation) and `sweep_prepare_run.py` (per-run preparation) for better separation of concerns
- **Centralized sweep coordination**: Implemented database-backed sweep coordination API for atomic run ID generation, preventing race conditions in parallel worker scenarios
- **Process-specific configuration**: Added support for process-specific file naming (`dist_{process_id}.yaml`, `train_config_overrides_{process_id}.yaml`) to avoid file conflicts between parallel workers
- **Updated parameter naming**: Standardized on `sweep_name` parameter across all sweep utilities and configurations

### Backend URL Configuration Support  
- **Configurable sweep coordination backend**: Added support for configuring backend URL through `stats_server_uri` configuration parameter
- **Clean API design**: Updated `metta_client_utils` functions to accept `backend_url` parameter without tight config coupling
- **Authentication integration**: Leverages existing machine token system from `~/.metta/observatory_tokens.yaml`
- **Development setup**: Configured sweep jobs to use `localhost:8000` for local development instead of production API
- **Backward compatibility**: Maintains localhost fallback for existing workflows

### Distributed Training Port Conflict Fix
- **Conditional distributed training**: Modified `devops/train.sh` to only use `torchrun` when truly needed (multi-node OR multi-GPU scenarios)
- **Eliminates unnecessary overhead**: Single GPU/single node training now uses direct Python execution, avoiding distributed machinery
- **Fixes parallel sweep worker crashes**: Resolves `EADDRINUSE` port conflicts when multiple sweep workers run simultaneously on same machine
- **Maintains distributed capabilities**: Multi-GPU and multi-node training continues to work with full distributed coordination

### Comprehensive Testing
- **Integration tests**: Added comprehensive test coverage for centralized sweep coordination, atomic run ID generation, process-specific file handling, and pipeline integration
- **Script-level testing**: Added tests for sweep script argument parsing, distributed config generation, and environment isolation
- **Configuration loading tests**: Verified trainer override application and configuration merging behavior

### Configuration Updates
- **Trainer override loading**: Fixed `load_train_job_config_with_overrides` to support process-specific override files while maintaining backward compatibility
- **Sweep configuration updates**: Updated sweep job configurations to work with new centralized coordination system
- **Shell script improvements**: Updated sweep rollout scripts to work with refactored system and new parameter naming

### Impact
- **Enables parallel sweep execution**: Multiple sweep workers can now run simultaneously without conflicts
- **Improves development workflow**: Local development uses local backend, production uses production API automatically  
- **Reduces training overhead**: Single-process scenarios no longer have unnecessary distributed training setup
- **Better reliability**: Atomic run ID generation prevents race conditions in distributed sweep scenarios
- **Enhanced testing**: Comprehensive test coverage ensures system reliability and catches regressions

**Files changed:** 12 files across sweep utilities, configuration, training scripts, and test suites

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1210840294824236)

### CODE REVIEW

**Comment by akrbc9** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1578?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1578** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1578?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1578?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1559** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1559?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1375](https://github.com/Metta-AI/metta/pull/1375) <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1375?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1554** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/1554?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: devops/train.sh [hunk 1]
```diff
@@ -45,19 +45,29 @@ echo "  - Master port: $MASTER_PORT"
 echo "  - Node index: $NODE_INDEX"
 echo "  - Arguments: $args"
 
-echo "[INFO] Starting distributed training..."
+echo "[INFO] Starting training..."
 
 set +e
```
Comment (akrbc9) [inline devops/train.sh:50]: Can remove if you don't want the extra complexity from the patch.

File: devops/sweep_rollout.sh [hunk 2]
```diff
@@ -12,15 +12,21 @@ if [ -z "$sweep_name" ]; then
   exit 1
 fi
```
Comment (akrbc9) [inline devops/sweep_rollout.sh:14]: This does need to stay because DATA_DIR is typically an S3 bucket for Sky jobs (if I'm not wrong), so different sweep processes may try to grab the same dist_file_cfg without it

File: app_backend/src/metta/app_backend/sweep_client.py [hunk 3]
```diff
@@ -61,11 +61,14 @@ def get_next_run_id(self, sweep_name: str) -> str:
         return data.run_id
 
 
-# Example usage for generate_run_id_for_sweep():
+# DEPRECATED: Use get_next_run_id_from_metta from metta.sweep.metta_client_utils instead
```
Comment (daveey) [inline app_backend/src/metta/app_backend/sweep_client.py:64]: can we remove?

File: configs/sweep_job.yaml [hunk 4]
```diff
@@ -9,6 +9,10 @@ defaults:
 
 cmd: sweep
 sweep_name: ???  # Will be set via command line override
+sweep_process_id: localhost  # Process-specific ID for parallel workers, overridden via CLI
```
Comment (daveey) [inline configs/sweep_job.yaml:12]: nope, we dont run parallel workers on one machine

File: configs/sweep_job.yaml [hunk 5]
```diff
@@ -9,6 +9,10 @@ defaults:
...
+
+# Override stats_server_uri from common.yaml to use localhost for sweep coordination
+stats_server_uri: http://localhost:8000
```
Comment (daveey) [inline configs/sweep_job.yaml:15]: define sweep_server_uri, don't rely on stats_server

File: devops/sweep.sh [hunk 6]
```diff
@@ -2,6 +2,10 @@
 # sweep.sh - Continuous sweep execution with retry logic
 set -e
 
+# Generate unique process ID for this sweep worker
+SWEEP_PROCESS_ID=$(python -c "import uuid; print(uuid.uuid4().hex[:8])")
```
Comment (daveey) [inline devops/sweep.sh:6]: nope

File: devops/sweep.sh [hunk 7]
```diff
@@ -17,9 +21,10 @@ fi
 # Extract sweep name from run
 sweep_name=$(echo "$args" | grep -E -o '(^|[[:space:]])run=[^ ]*' | sed 's/.*run=//')
 
-
 # Replace run=<name> with sweep_name=<name> - handle both start of string and after space
 args_for_rollout=$(echo "$args" | sed 's/^run=/sweep_name=/' | sed 's/ run=/ sweep_name=/g')
+# Add process ID to arguments (not as environment variable)
```
Comment (daveey) [inline devops/sweep.sh:26]: nope

File: devops/sweep_rollout.sh [hunk 8]
```diff
@@ -12,15 +12,21 @@ if [ -z "$sweep_name" ]; then
...
 
+# Extract sweep process ID
+sweep_process_id=$(echo "$args" | grep -o 'sweep_process_id=[^ ]*' | sed 's/sweep_process_id=//' || echo "")
```
Comment (daveey) [inline devops/sweep_rollout.sh:16]: nope. please just remove all the multi-process support stuff

File: devops/train.sh [hunk 9]
```diff
@@ -45,19 +45,29 @@ echo "  - Master port: $MASTER_PORT"
...
+    --nproc-per-node=$NUM_GPUS \
+    --master-addr=$MASTER_ADDR \
+    --master-port=$MASTER_PORT \
+    --node-rank=$NODE_INDEX \
+    tools/train.py \
+    trainer.num_workers=null \
+    wandb.enabled=true \
+    $args
+else
+  echo "[INFO] Using single-process training (no distributed overhead)"
```
Comment (daveey) [inline devops/train.sh:65]: nope. let' always use torchrun. there is no overhead and 1 code path is easier to debug

File: metta/sweep/metta_client_utils.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
+from metta.app_backend.routes.sweep_routes import SweepCreateResponse
+from metta.app_backend.sweep_client import SweepClient
+from metta.common.util.stats_client_cfg import get_machine_token
+
+
+def _get_sweep_client(backend_url: str | None = None) -> SweepClient:
```
Comment (daveey) [inline metta/sweep/metta_client_utils.py:6]: put into 
```
SweepClient.__new__()
```

File: metta/sweep/metta_client_utils.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    if backend_url is None:
+        backend_url = "http://localhost:8000"
+
+    # Get machine token for authentication (same pattern as stats client)
+    auth_token = get_machine_token(backend_url)
+
+    return SweepClient(base_url=backend_url, auth_token=auth_token)
+
+
+def get_sweep_id_from_metta(sweep_name: str, backend_url: str | None = None) -> str | None:
```
Comment (daveey) [inline metta/sweep/metta_client_utils.py:25]: put this into CogwebClient:
  sweep_id()
  sweep_next_run_id()
...

File: metta/sweep/wandb_utils.py [hunk 12]
```diff
@@ -111,6 +111,10 @@ def sweep_id_from_name(project: str, entity: str, name: str) -> str | None:
 
 
 def generate_run_id_for_sweep(sweep_id: str, sweep_names_dir: str) -> str:
+    """DEPRECATED: Use get_next_run_id_from_metta() from metta.sweep.metta_client_utils instead.
```
Comment (daveey) [inline metta/sweep/wandb_utils.py:114]: remove func

File: tools/sweep_prepare_run.py [hunk 13]
```diff
@@ -41,8 +41,9 @@ def setup_next_run(cfg: DictConfig | ListConfig, logger: Logger) -> str:
     sweep_metadata = OmegaConf.load(os.path.join(cfg.sweep_dir, "metadata.yaml"))
 
     # Generate a new run ID for the sweep, e.g. "simple_sweep.r.0"
-    # TODO: Use sweep_id instead of sweep_path, currently very confusing.
-    run_id = generate_run_id_for_sweep(sweep_metadata.wandb_path, cfg.runs_dir)
+    # Use centralized database for atomic run ID generation
+    backend_url = cfg.stats_server_uri if hasattr(cfg, "stats_server_uri") else None
```
Comment (daveey) [inline tools/sweep_prepare_run.py:45]: should always be there, why the check?

File: tools/sweep_prepare_run.py [hunk 14]
```diff
@@ -83,6 +84,9 @@ def init_run():
             assert isinstance(sweep_job_container, dict), "sweep_job must be a dictionary structure"
             sweep_job_copy = DictConfig(sweep_job_container)
             apply_protein_suggestion(sweep_job_copy, clean_suggestion)
+            # NOTE: Overrides are expected to be in the run_dir as `train_config_overrides.yaml`.
```
Comment (daveey) [inline tools/sweep_prepare_run.py:87]: remove

**Review by daveey — CHANGES_REQUESTED** [review]

File: tools/sweep_prepare_run.py (hunk 14 above)
Comment (akrbc9) [reply tools/sweep_prepare_run.py:87]: are you talking about the assert statement?

File: cogweb/cogweb_client.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
+"""Unified client for all Cogweb backend services."""
+
+from typing import Optional
+
+from metta.app_backend.sweep_client import SweepClient
+
+
+class CogwebClient:
```
Comment (Barraketh) [inline cogweb/cogweb_client.py:8]: i'm confused.  What is the point of this file? Why are we not just using the sweep client directly?

File: app_backend/src/metta/app_backend/sweep_client.py [hunk 16]
```diff
@@ -16,12 +16,18 @@ class SweepClient:
 
     def __init__(self, base_url: str = "http://localhost:8000", auth_token: Optional[str] = None):
         """
-        Initialize the sweep client.
+        Initialize the sweep client with automatic authentication if no token provided.
 
         Args:
             base_url: Base URL of the API server
-            auth_token: Authentication token (machine token or user session)
+            auth_token: Authentication token. If None, will attempt to get machine token.
```
Comment (Barraketh) [inline app_backend/src/metta/app_backend/sweep_client.py:23]: I don't like this dependency - I don't really want app_backend to have dependencies on the rest of the repo, and in particular on hydra.  I would keep the SweepClient as is, and then if you need to add a 'get_sweep_client' next to 'get_stats_client'

File: cogweb/cogweb_client.py (hunk 15 above)
Comment (daveey) [inline cogweb/cogweb_client.py:8]: I was thinking 

```
CogwebClient(cogweb_config): 
  stats_logger(): StatsClient
  sweep_manager(): SweepManager
  policy_store(): PolicyStore
```

This would let us define a small number of CogwebConfig, one for softmax, one for local, etc...

I guess we could just create them directly as well, I don't have a strong preference for now

**Review by daveey — CHANGES_REQUESTED** [review]
Looking better

File: configs/sweep_job.yaml [hunk 17]
```diff
@@ -10,6 +10,9 @@ defaults:
 cmd: sweep
 sweep_name: ???  # Will be set via command line override
 
+# Override stats_server_uri from common.yaml to use localhost for sweep coordination
```
[the remark this answers is not in the document]
Comment (akrbc9) [reply configs/sweep_job.yaml:13]: Leftover comment from when this was stats_server_uri. Removed upstack in #1591

**Comment by akrbc9** [comment]
refactored cogweb client. Will be easy to make it config dependent when necessary.

File: configs/sweep_job.yaml [hunk 18]
```diff
@@ -10,6 +10,9 @@ defaults:
...
+sweep_server_uri: http://localhost:8000
```
Comment (Barraketh) [inline configs/sweep_job.yaml:14]: Why are we doing this?

**Review by relh — APPROVED** [review]
